pyomo/common/collections/_hasher.py

```diff
@@ -0,0 +1,66 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2025
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from collections import defaultdict
+
+
+class Hasher(defaultdict):
+    """
+    Hasher is a dictionary-like class that provides a consistent hashing for Python objects.
+
+    This class allows registration of custom hashing functions for specific types. When an object is passed to the Hasher, it determines the appropriate hashing strategy based on the object's type:
+    - If a custom hashing function is registered for the type, it is used.
+    - If the object is natively hashable, the default hash is used.
+    - If the object is unhashable, the object's `id()` is used as a fallback.
+
+    The Hasher also includes a special handling for tuples by recursively applying the appropriate hashing strategy to each element within the tuple.
+    """
```

Comment:
Some minor edits; the important change is to wrap the strings to 88 characters (for consistenct with the Black line wrapping for code):
```suggestion
    """Dispatch table for generating "universal" hashing of all Python objects.

    This class manages a dispatch table for providing hash functions for all Python 
    types.  When an object is passed to the Hasher, it determines the appropriate
    hashing strategy based on the object's type:
    - If a custom hashing function is registered for the type, it is used.
    - If the object is natively hashable, the default hash is used.
    - If the object is unhashable, the object's `id()` is used as a fallback.

    The Hasher also includes a special handling for tuples by recursively applying the
    appropriate hashing strategy to each element within the tuple.
    """
```